src/main/WikiVectors.java

```diff
@@ -0,0 +1,165 @@
+                }
+                if (vectors.size() % 10000 == 0) {
+                    System.out.print("loaded " + vectors.size() + "\r");
+                    System.out.flush();
+                }
+            }
+        }
+        System.out.println("loaded " + vectors.size());
+    }
+
+    void computeVectors(String lineDocFile, String outputFile) throws IOException {
+        int count = 0;
+        CharsetDecoder dec=StandardCharsets.UTF_8.newDecoder()
+            .onMalformedInput(CodingErrorAction.REPLACE); // replace invalid input with the UTF8 replacement character
+        try (OutputStream out = Files.newOutputStream(Paths.get(outputFile));
+             Reader r = Channels.newReader(FileChannel.open(Paths.get(lineDocFile)), dec, -1);
+             BufferedReader in = new BufferedReader(r)) {
+            String lineDoc;
+            byte[] buffer = new byte[dimension * Float.BYTES];
+            ByteBuffer bbuf = ByteBuffer.wrap(buffer)
+                .order(ByteOrder.LITTLE_ENDIAN);
+            FloatBuffer fbuf = bbuf.asFloatBuffer();
+            while ((lineDoc = in.readLine()) != null) {
+                float[] dvec = computeDocVector(lineDoc);
+                fbuf.position(0);
+                fbuf.put(dvec);
+                out.write(buffer);
+                if (++count % 10000 == 0) {
+                    System.out.print("wrote " + count + "\r");
+                    System.out.flush();
+                }
+            }
+            System.out.println("wrote " + count);
+            System.out.println("Token Coverage = " + (100 * tokenCount / (tokenCount + missingTokenCount)) + "%");
+        } catch (IOException e) {
+            System.err.println("An error occurred on line " + (count + 1));
+            throw e;
+        }
+    }
+
+    void vectorAdd(float[] x, float[] y) {
+        assert x.length == y.length;
+        for (int i = 0; i < x.length; i++) {
+            x[i] += y[i];
+        }
+    }
+
+    void vectorDiv(float[] v, float x) {
+        for (int i = 0; i < v.length; i++) {
+            v[i] /= x;
+        }
+    }
+
+    float[] computeDocVector(String doc) {
+        float[] dvec = new float[dimension];
+        int count = 0;
+        for (String token : tokenize(doc)) {
+            float[] tvec = vectors.get(token);
+            if (tvec != null) {
+                vectorAdd(dvec, tvec);
+                count++;
+            } else {
+                missingTokenCount++;
+            }
+        }
+        tokenCount += count;
+        vectorDiv(dvec, count);
+        return dvec;
+    }
+
+    // tokenize on white space, most ascii punctuation, preserving -_,., then lower case.  Not very
+    // sophisticated, but enough for performance testing on English text. Should we use StandardTokenizer?
+    Iterable<String> tokenize(String document) {
+        List<String> tokens = new ArrayList<>();
+        for (String t : document.split("[\\]\\[\\\\:\"'?/<> \t~`!@#$%^&*\\(\\)+={}]+")) {
```

Comment:
Look at that regexp!  This is trying to approximate `StandardAnalyzer`?  Or, does it really matter?  This is just used to aggregate per-token vectors into a single vector for the document, and is an offline process (before benchmark runs its `Indexer`)?

Reply:
It's actually trying to mimic whatever GloVe was produced with, which I think is this https://nlp.stanford.edu/software/tokenizer.shtml. However, for the purpose of performance evaluation I think "close enough" is fine - we just need to ensure a reasonable distribution of document vectors. We might also want consistent behavior across document and query vectors, but since we don't have any query vectors (yet!) that's trivial.